src/env/double_write.rs

```diff
@@ -0,0 +1,722 @@
+// Copyright (c) 2017-present, PingCAP, Inc. Licensed under Apache-2.0.
+
+use crate::file_pipe_log::log_file::build_file_reader;
+use crate::file_pipe_log::pipe_builder::FileName;
+use crate::file_pipe_log::reader::LogItemBatchFileReader;
+use crate::file_pipe_log::FileNameExt;
+use crate::internals::parse_reserved_file_name;
+use crate::internals::FileId;
+use crate::internals::LogQueue;
+use crate::{Error, Result};
+use crossbeam::channel::unbounded;
+use crossbeam::channel::Sender;
+use fail::fail_point;
+use log::{info, warn};
+use std::fs;
+use std::io::{Read, Result as IoResult, Seek, SeekFrom, Write};
+use std::path::Path;
+use std::path::PathBuf;
+use std::sync::atomic::AtomicU64;
+use std::sync::atomic::Ordering;
+use std::sync::Arc;
+use std::sync::RwLock;
+use std::thread;
+
+use crate::env::{FileSystem, Handle, Permission, WriteExt};
+use futures::channel::oneshot;
+use futures::executor::block_on;
+use futures::select;
+
+use either::Either;
+
+type Callback<T> = Box<dyn FnOnce(IoResult<Option<T>>) + Send>;
+
+#[derive(PartialEq)]
+enum Task {
+    Create(PathBuf),
+    Open {
+        path: PathBuf,
+        perm: Permission,
+    },
+    Delete(PathBuf),
+    Rename {
+        src_path: PathBuf,
+        dst_path: PathBuf,
+    },
+    Stop,
+}
+
+#[derive(Default)]
+struct Files {
+    prefix: PathBuf,
+    append_file: Vec<FileName>,
+    rewrite_file: Vec<FileName>,
+    recycled_file: Vec<FileName>,
+}
+
+fn replace_path(path: &Path, from: &Path, to: &Path) -> PathBuf {
+    if let Ok(file) = path.strip_prefix(from) {
+        to.to_path_buf().join(file)
+    } else {
+        panic!("Invalid path: {:?}", path);
+    }
+}
+
+pub struct HedgedFileSystem<F: FileSystem> {
+    base: Arc<F>,
+
+    path1: PathBuf,
+    path2: PathBuf,
+    disk1: Sender<(Task, Callback<F::Handle>)>,
+    disk2: Sender<(Task, Callback<F::Handle>)>,
+
+    handle1: Option<thread::JoinHandle<()>>,
+    handle2: Option<thread::JoinHandle<()>>,
+}
+
+// TODO: read both dir at recovery, maybe no need? cause operations are to both
+// disks TODO: consider encryption
+
+impl<F: FileSystem> HedgedFileSystem<F> {
+    pub fn new(base: Arc<F>, path1: PathBuf, path2: PathBuf) -> Self {
+        let (tx1, rx1) = unbounded::<(Task, Callback<F::Handle>)>();
+        let (tx2, rx2) = unbounded::<(Task, Callback<F::Handle>)>();
+        let fs1 = base.clone();
+        let handle1 = thread::spawn(|| {
+            for (task, cb) in rx1 {
+                if task == Task::Stop {
+                    break;
+                }
+                fail_point!("double_write::thread1");
+                let res = Self::process(&fs1, task);
+                cb(res);
+            }
+        });
+        let fs2 = base.clone();
+        let handle2 = thread::spawn(|| {
+            for (task, cb) in rx2 {
+                if task == Task::Stop {
+                    break;
+                }
+                let res = Self::process(&fs2, task);
```

Comment:
why doesn't set a fail_point of this handle?

Reply:
It's used to slow one disk not both